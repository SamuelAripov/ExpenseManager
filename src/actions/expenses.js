import database from '../firebase/firebase';
import { get } from 'http';

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt }

        return database.ref(`users/${uid}/expenses`).push(expense).then(({ ref }) => {
            console.log(ref)
            dispatch(addExpense({
                id: expense.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
})
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})
export const clearExpenses = () => ({
    type: 'CLEAR_EXPENSES'
})
export const startSetExpense = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        dispatch(clearExpenses())
        return database
            .ref(`users/${uid}/expenses/`)
            .once('value')
            .then((snapshot) => {
                snapshot.forEach((childSnapshot) => {
                    dispatch(addExpense({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    }))
                    console.log('Official Data', {
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    })
                })
            }).catch(e => console.log(e))
    }
}
export const startRemoveExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }))
            })
    }
}
export const startEditExpense = (id, update) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update({
                ...update
            })
            .then(() => {
                dispatch(editExpense(id, update))
            })
    }
}

