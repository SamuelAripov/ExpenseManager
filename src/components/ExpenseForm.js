import React from 'react'
import DatePicker from 'react-datepicker'
export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? props.expense.createdAt : new Date().valueOf(),
            calendarFocused: false
        }
    }
    
    onDescripitionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onTextFormChange = (e) => {
        const textValue = e.target.value
        this.setState(() => ({ note: textValue }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        this.props.onSubmit({
            description: this.state.description,
            amount: this.state.amount,
            createdAt: this.state.createdAt,
            note: this.state.note
        })
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescripitionChange}
                        required
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        required
                    />
                    <br />
                    <DatePicker
                        selected={this.state.createdAt}
                        onChange={this.onDateChange}
                        dateFormat="MMMM d, yyyy"
                        onChangeRaw={(e) => e.preventDefault()}
                    />
                    <textarea
                        placeholder="Add a note for you expense (optional)"
                        value={this.state.note}
                        onChange={this.onTextFormChange}
                    >
                    </textarea>
                    <button>Submit</button>
                </form>
            </div>
        )
    }
}