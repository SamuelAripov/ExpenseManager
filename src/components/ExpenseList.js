import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getFilteredItems from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div className='content-container'>
        {props.expenses.length !== 0 ? (
            <div className='list-header'>
                <div className='show-for-mobile'>Expenses</div>
                <div className='show-for-desktop'>Expense</div>
                <div className='show-for-desktop'>Amount</div>
            </div>
        ) : <span></span>}
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense} />
                    })
                )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: getFilteredItems(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)