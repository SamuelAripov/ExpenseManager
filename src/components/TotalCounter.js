import React from 'react'
import { connect } from 'react-redux'
import getFilteredItems from '../selectors/expenses'
import numeral from 'numeral'
export class TotalCounter extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            sum: 0
        }
        if (this.props.expenses) {
            let sum = 0
            this.props.expenses.map((expense) => {
                sum += parseFloat(expense.amount)
            })
            this.state = {
                sum
            }
        }
    }
    render() {
        return (
            <div>
                {this.props.expenses ? <h1>Viewing {this.props.expenses.length} expense(s) with the total of {numeral(this.state.sum).format('$0,0.00')}</h1> : <h1></h1>}
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        expenses: getFilteredItems(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(TotalCounter)