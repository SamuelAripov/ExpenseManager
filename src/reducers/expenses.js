// Expenses Reducer
import database from '../firebase/firebase'

export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': {
            console.log(action.expense)
            return [
                ...state,
                action.expense
            ]
        }
        case 'REMOVE_EXPENSE': {
            return state.filter((expense) => expense.id !== action.id)
        }
        case 'EDIT_EXPENSE': {
            return state.map(expense => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else
                    return expense
            })
        }
        case 'SET_EXPENSES': {
            return action.expenses
        }
        case 'CLEAR_EXPENSES': {
            return []
        }
        default:
            return state;
    }
}