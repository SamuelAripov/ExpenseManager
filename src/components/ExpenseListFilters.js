import React from 'react'
import { connect } from 'react-redux'
import DatePicker from 'react-datepicker'
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'
export class ExpenseListFilters extends React.Component {
    constructor(props){
        super(props)
        this.state={
            text: this.props.filters ? this.props.filters.text : '',
            sortBy: this.props.filters ? this.props.filters.sortBy : 'date',
            startDate: this.props.filters ? this.props.filters.startDate : new Date().valueOf(),
            endDate: this.props.filters ? this.props.filters.endDate : new Date().valueOf()
        }
    }
    render() {
        return (
            <div>
                <label>Search: <input type='text' value={this.props.filters.text} onChange={(e) => {this.props.setTextFilter(e.target.value)}}></input></label>
                <select onChange={(e) => {
                    e.target.value === 'date' ? this.props.sortByDate('Date') : this.props.sortByAmount('AmountRising')
                }}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <span>Start Date</span>
                <DatePicker
                    selected={this.props.filters.startDate}
                    onChange={(e) => this.props.setStartDate(e.valueOf())}
                    dateFormat="MMMM d, yyyy"
                    onChangeRaw={(e) => e.preventDefault()}
                />
                <span>End Date</span>
                <DatePicker
                    selected={this.props.filters.endDate}
                    onChange={(e) => this.props.setEndDate(e.valueOf())}
                    dateFormat="MMMM d, yyyy"
                    onChangeRaw={(e) => e.preventDefault()}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: () => dispatch(sortByAmount()),
        setStartDate: (startDate) => dispatch(setStartDate(startDate)),
        setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)