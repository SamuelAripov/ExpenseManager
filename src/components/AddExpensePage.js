import React from 'react'
import { startAddExpense } from '../actions/expenses'
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'

export class AddExpensePage extends React.Component {
    render() {
        return (
            <div className='content-container'>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={(expense) => {
                        this.props.startAddExpense(expense)
                        this.props.history.push('/dashboard')
                        // location.reload()
                    }}
                />
            </div>)
    }
}
const mapDispatchToProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    }
}
export default connect(undefined, mapDispatchToProps)(AddExpensePage)