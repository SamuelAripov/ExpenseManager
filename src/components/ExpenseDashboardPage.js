import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import TotalCounter from '../components/TotalCounter'

const ExpenseDashboardPage = () => (
    <div>
        <TotalCounter/>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
)

export default ExpenseDashboardPage