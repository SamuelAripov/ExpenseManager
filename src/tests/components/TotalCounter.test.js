import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import { TotalCounter } from '../../components/TotalCounter'


test('should render message without expenses', () => {
    const wrapper = shallow(<TotalCounter expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render message with one expense', () => {
    const wrapper = shallow(<TotalCounter expenses={[expenses[1]]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render message with multiple expenses', () => {
    const wrapper = shallow(<TotalCounter expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})