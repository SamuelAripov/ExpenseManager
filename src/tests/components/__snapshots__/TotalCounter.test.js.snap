// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render message with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     expense(s) with the total of 
    $6.00
  </h1>
</div>
`;

exports[`should render message with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense(s) with the total of 
    $2.00
  </h1>
</div>
`;

exports[`should render message without expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     expense(s) with the total of 
    $0.00
  </h1>
</div>
`;
